Feature - 네트워크 세팅

yangsubinn opened the pull request:
## 관련 이슈
<!--
이슈 번호 #000 작성  
ex) close #1 
--> 
close #118 

## 내용
<!--
로직 설명  
--> 

**1. BaseAPI, UserAPI 파일 생성**
- 라우터별로 정의한 APIType 생성했습니다.
- BaseAPI에서는 그 외의 API들이 상속받아 apiType을 넣어주고, apiType에 따라 baseURL 뒤에 기본 path 추가하게 했습니다.
- 각 라우터별 API enum에서는 각각 api 케이스를 넣어주고 BaseAPI 상속받은 API 안에 각각 다른 부분을 서술해주면 됩니다
- 확인용으로 회원가입 테스트해보려고 UserAPI만 우선 생성했습니다
- 네트워크 흐름 확정되면 나머지 API 파일도 추가해서 올리겠습니다.

**2. BaseService 생성**
- 각 라우터별 API(ex. UserAPI)를 받아서 그 안에 정의된 api targetType에 따라 네트워크 요청 보낼 수 있도록 requestObject 함수 추가했습니다.
- requestObjectInCombine도 넣어두긴 했는데, 전체적으로 combine 흐름을 사용하지 않다보니 필요가 없는 것 같아서 삭제하게될 것 같습니다.

**3. 라우터별 Service**
- 각 라우터별로 존재하는 api 요청을 프로토콜로 관리하는게 실수를 줄일 수 있을 것 같아 별도로 정의했습니다.
- 프로토콜 내부에 UserAPI에 해당하는 요청을 전부 정의해두고 그 아래 DefaultUserService(BaseService<UserAPI>에서 요청을 보내는 것으로 생각했습니다

**4. 뷰 별 Repository**
- SignUpService 내부에서 바로 BaseService<>에 접근해서 요청을 보내기가 복잡해서 기존에 있던 MockUp 구조체를 Repository로 사용했습니다.
- 사용하게 된다면 별도로 폴더, 파일 분리해서 사용하면 좋을 것 같습니다 

**5. 회원가입 연결 시도**
- 회원가입 테스트는 해보려고 했는데 뭔가 한번 하면 그 이후 기회가 없을 것 같아서 .. 주석처리해두었습니다
- SignUpFeature에서 SignUpRequestModel 만들어서 넣도록 했고, 
- nickname, profileImageUrl만 연결해두었습니다. (platform, fcmToken, 소셜로그인 token 연결 X)

**6. Baggle.config 파일에 BASE_URL 추가**


### 이미지, 영상
<!--
필요시 스크린샷 첨부
--> 

## 리뷰어가 확인할 사항
<!--
중점적으로 봐주면 좋을 사항  
ex) ㅇㅇㅇㅇ하려고 Combine 사용했는데 제대로 사용하고 있는건가요?
--> 

- 회원가입 부분은 아직 완벽하게 연결이 안됐습니다. 여기서는 흐름만 연결해보고 따로 이슈 파서 해야할 것 같습니다.
    - networkService에 요청하는 부분 주석처리해두었습니다
- 지금 구현된 것처럼 뷰별 Repository와 라우터별 Service 파일이 있으면 어떨까요?
    - 라우터별 Service : BaseService 통해서 실제 통신 요청
    - 뷰별 Repository : 라우터별 service 파일에 요청 후 entity 데이터 받아서 toDomain 변환 후 뷰별 Service에 전달
    - 폴더링은 Core > Common > Network 안에 넣으면 어떨까 싶었는데 의견 있으면 말해주세요
- combine을 넣자니 뭔가 더 복잡해질 것 같아서 combine 사용하지 않고 했습니다
- 회원가입이 아직 entity 확인이 안돼서 이걸 테스트를 해봐야될지 어찌할지 모르겄네요...... 
- 서버분들께 더미유저를 요청해서 get 통신 먼저 해보는게 나을 것 같습니다


## 기타
<!--
레퍼런스 혹은 패키지 설치 등
-->

yangsubinn committed:
Feat #118 - config, info.plist BaseAPI 추가

yangsubinn committed:
Feat #118 - BaseAPI, UserAPI 추가

yangsubinn committed:
Chore #118 - UserAPI 주석 추가

yangsubinn committed:
Merge: develop pull

yangsubinn committed:
Feat #118 - BaseService 추가
- DefaultAlamofireManager, SignUpEntity 추가
- EntityContainer 디코딩 추가

yangsubinn committed:
Feat #118 - SignUpService 연결
- SighUpService와 AuthService 분리
- mockUpSignUpService -> repository
- SignUpFeature에서 string부분 requestModel로 수정

yangsubinn committed:
Chore #118 - 파일명 수정

yangsubinn committed:
Merge: develop pull
Merge branch 'develop' of https://github.com/dnd-side-project/dnd-9th-2-ios into feature/api-signup

never-better commented on Baggle/Baggle/Core/Models/Network/EntityContainer.swift:
```diff
@@ -7,8 +7,23 @@
+    }
+}
+
+extension EntityContainer: Decodable where T: Decodable {
```
p3: extension으로 분리하고 CodingKeys를 사용한 이유가 있나요?? 특별한 이유가 없으면 아래와 같이 작성할 수 있습니다.


``` swift
struct EntityContainer<T>: Decodable where T: Decodable{
    let status: Int
    let message: String
    let data: T
}
```

never-better commented on Baggle/Baggle/Core/Models/Network/SignUp/SignUpEntity.swift:
```diff
@@ -0,0 +1,27 @@
+    let accessToken: String
+    let refreshToken: String
+    let platform: String
+    let userId: Int?
```
p3: 컨벤션에 맞춰서 아이디는 시작할때는 id, 그 외에는 다 ID로 통일하면 좋을 것 같아요.

let id: String
let userID: String

never-better replied:
Q: 지금 userID가 추가 안 되어있어서 옵셔널인거죠??

yangsubinn replied:
명세서 상에서 데이터가 meetingId, feedId 요런 식으로 되어있어서 userId로 해뒀습니다..! 컨벤션에 맞추는게 더 나을까요?
네넵 근데 userID 추가해주신다고 해서 옵셔널 빼버리겠습니다

never-better replied:
Swift 내부에서는 ID로 통일하는게 좋을 것 같아요

never-better commented on Baggle/Baggle/Features/Login/SignUp/Main/SignUpFeature.swift:
```diff
@@ -150,11 +158,13 @@ struct SignUpFeature: ReducerProtocol {
                     await send(.imageLoading)
-
+                    
                     if let profileImage = try? await photoPickerItem?.loadTransferable(
```
p2: Profile Image Model을 수정하면 더 깔끔하게 할 수 있을 것 같아요.

```swift

struct ProfileImageModel: Transferable {
    let image: UIImage

    static var transferRepresentation: some TransferRepresentation {
        DataRepresentation(importedContentType: .image) { data in
            guard let uiImage = UIImage(data: data) else {
                throw TransferError.importFailed
            }
            return ProfileImageModel(image: uiImage)
        }
    }
}
```

never-better replied:
그리고 feature State에서 UIImage를 들고 있다가 버튼 누르면 Data로 변환해주면 될 것 같아요.

```swift
// 버튼 누를 때
let imageData = image.jpegData(compressionQuality: 0.8)
```
```swift
// Profile Image View
case .success(let image):
           Image(uiImage: image)
                .resizable()

// Album Image State
enum AlbumImageState: Equatable {
    case success(UIImage)
    // 그 외...
}
```

never-better replied:
제가 나중에 하려고 미뤄뒀는데 괜히 할 일을 만들었네요... 😇

never-better reviewed (commented):
제가 Demo App으로 만들어본 건데 Combine -> async/await 해서 TCA에 적용했습니다. 이렇게 하면 클로저랑 Combine 없이 async, await으로 사용할 수 있을 것 같은데 어떠신가요?

```swift
import Foundation

import Moya

extension MoyaProvider {
    func request(_ target: Target) async throws -> Data {
        return try await withCheckedThrowingContinuation({ continuation in
            self.request(target) { result in
                switch result {
                    
                case .success(let response):
                    continuation.resume(returning: response.data)
                    
                case .failure(let error):
                    continuation.resume(throwing: error)
                }
            }
        })
    }
}
```

yangsubinn committed:
Fix #118 - 코드리뷰 반영 수정
- BaseService 내 request 함수 추가, 기존 requestObjectInCombine 삭제
- UserService 파일 삭제
- Image 변환 ProfileImageModel 수정
- NetworkError 추가

yangsubinn committed:
Chore #118 - 주석 추가

yangsubinn committed:
Merge: develop pull

yangsubinn commented:
> 제가 Demo App으로 만들어본 건데 Combine -> async/await 해서 TCA에 적용했습니다. 이렇게 하면 클로저랑 Combine 없이 async, await으로 사용할 수 있을 것 같은데 어떠신가요?
> 
> ```swift
> import Foundation
> 
> import Moya
> 
> extension MoyaProvider {
>     func request(_ target: Target) async throws -> Data {
>         return try await withCheckedThrowingContinuation({ continuation in
>             self.request(target) { result in
>                 switch result {
>                     
>                 case .success(let response):
>                     continuation.resume(returning: response.data)
>                     
>                 case .failure(let error):
>                     continuation.resume(throwing: error)
>                 }
>             }
>         })
>     }
> }
> ```

남겨주신 request 추가해서 기존에 라우터별 Service 파일 없애고 repository에서 아래처럼 바로 요청보낼 수 있도록 수정했습니다.
```
struct SignUpRepository {
    private let networkService = BaseService<UserAPI>()
    
    func fetchSignUp(requestModel: SignUpRequestModel) async throws -> SignUpServiceState {
        print("📍requestModel: \(requestModel)")
        
        do {
            let data: SignUpEntity = try await networkService.request(.signUp(requestModel: requestModel))
            print("data: \(data)")
            // userDefault 저장
            return .success
        } catch let error {
            print("SignUpRepository - error: \(error)")
            // error에 따라 분기처리
//            return .fail
            return .success
        }
    }
}
```

SignUpService -> SignUpRepository -> BaseService - request  요 흐름이 괜찮으시다면 네트워크 부분 우선 이렇게 하고 
다음 이슈에서 회원가입 네트워크 통신 작업 마저해보겠습니다!

yangsubinn committed:
Add #118 - API 파일 추가, HomeView 닉네임 text 수정

never-better reviewed (approved):
쉽지 않은 작업이었을텐데 수고하셨습니다! 덕분에 앞으로 네트워크 작업할 때 편하게 할 수 있을 것 같아요 ㅎㅎ

Q: Service -> Network Service를 바로 참고해도 될 것 같은데 Repository를 만든 이유가 있을까요??

yangsubinn commented:
> 쉽지 않은 작업이었을텐데 수고하셨습니다! 덕분에 앞으로 네트워크 작업할 때 편하게 할 수 있을 것 같아요 ㅎㅎ
> 
> Q: Service -> Network Service를 바로 참고해도 될 것 같은데 Repository를 만든 이유가 있을까요??

Service에서 바로 참고하려면 아래처럼 처리해줘야 하는데, 이게 여러개를 구현하다보면 조금 복잡해지기도 하고, 매번 BaseService<UserAPI>()로 불러와야 해서
repository에서 private let networkService = BaseService<UserAPI>()로 한번 정의해두고 여러 함수에서 사용 가능하게 하면서 Service에는 필요한 데이터 형태를 바로 받아볼 수 있도록 하면 좋을 것 같았습니다 !
```
extension SignUpService: DependencyKey {

    static var liveValue = Self { requestModel, token  in
        do {
            let data: SignUpEntity = try await BaseService<UserAPI>()
                .request(.signUp(requestModel: requestModel, token: token))
            // 로직 처리
            return SignUpServiceState.nicknameDuplicated
        } catch {
            return SignUpServiceState.fail
        }
    }
}
```

never-better commented:
확인했습니다!

yangsubinn committed:
Chore #118 - 프로비저닝 파일 수정

yangsubinn committed:
Fix #126 - build.yml 수정
